core/src/main/java/org/elasticsearch/index/Index.java

```diff
@@ -96,4 +110,37 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeString(name);
         out.writeString(uuid);
     }
+
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject();
+        builder.field(INDEX_NAME_KEY, name);
+        builder.field(INDEX_UUID_KEY, uuid);
+        return builder.endObject();
+    }
+
+    public static Index fromXContent(final XContentParser parser) throws IOException {
+        return INDEX_PARSER.parse(parser, new Builder()).build();
+    }
+
+    /**
+     * Builder for Index objects.  Used by ObjectParser instances only,
+     * otherwise just call the constructor directly.
+     */
+    final public static class Builder {
```

Comment:
The issue is, `IndexGraveyard.Tombstone` uses it for its own ObjectParser (because it must parse the `Index` using its object parser, which requires a Builder so it can remain immutable).  The alternative is to implement `Tombstone#fromXContent` using the bare bones `XContentParser` instead of the `ObjectParser`.